### PR TITLE
Mobifies mice

### PR DESCRIPTION
**Pull request by Bartimeus973**
## About the PR 
Turns all obj/critter/mice into mob critters. Also fixes some code that should allow them to spawn with semi-random colors.
Also adds some code to instantly kill mice who step on mousetraps. No it doesn't work on Remy. Yes it works on mentormice, but considering mentormice probably shouldnt be running around the place this should be fun for lighthearted ways to exit the round after mentoring someone. Admin mice are "unaffected".

## Why's this needed? 
Part of the effort to turn all obj/critters into mob critters. Also allows you to supplex/hug remy.

## Changelog

```changelog
(u)Bartimeus973
(+)Turns all mouse into mobs. Go hug Remy today!
```

### CODE REVIEW

**Comment by Bartimeus973** [comment]
Don't merge yet there appears to be a bug causing mentor/admin mice to turn into weak mice on ai ticks and im not sure why since is_npc is false. I'll check it out tomorrow.

**Comment by frawhst** [comment]
Overall this is a good change

File: code/mob/living/critter/ai/mouse.dm [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/datum/aiHolder/mouse_mad
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/mouse_mad, list(src))
+
+/datum/aiTask/prioritizer/critter/mouse_mad/New()
+	..()
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/critter/attack, list(holder, src))
+
```
[the remark this answers is not in the document]
Comment (Bartimeus973) [reply code/mob/living/critter/ai/mouse.dm]: They now eat stuff unless they have the option to try and eat your face off.

File: code/mob/living/critter/small_animal.dm [hunk 2]
```diff
@@ -136,23 +136,29 @@ ABSTRACT_TYPE(/mob/living/critter/small_animal)
...
-		fur_color = src.client?.preferences.AH.customization_first_color
-		eye_color = src.client?.preferences.AH.e_color
-		var/image/overlay = image('icons/misc/critter.dmi', "mouse_colorkey")
-		overlay.color = fur_color
-		src.UpdateOverlays(overlay, "hair")
-
-		var/image/overlay_eyes = image('icons/misc/critter.dmi', "mouse_eyes")
-		overlay_eyes.color = eye_color
-		src.UpdateOverlays(overlay_eyes, "eyes")
+		if (!istype(src, /mob/living/critter/small_animal/mouse/weak/mentor) && !istype(src, /mob/living/critter/small_animal/mouse/weak/mentor/admin) && !istype(src, /mob/living/critter/small_animal/mouse/remy))
```
[the remark this answers is not in the document]
Comment (Bartimeus973) [reply code/mob/living/critter/small_animal.dm]: Done

File: code/obj/item/mousetrap.dm [hunk 3]
```diff
@@ -315,20 +315,27 @@
...
-				"<span class='alert'><B>You accidentally trigger the mousetrap!</B></span>")
+		else if (istype(AM, /mob/living/critter/small_animal/mouse/weak/mentor/admin)) //The admin mouse fears not your puny attempt to squish it.
+			AM.visible_message("<span class='alert'>[src] blows up violently as soon as [AM] sets foot on it! [AM] looks amused at this poor attempt on it's life.</span>")
+			new/obj/effect/supplyexplosion(src.loc)
+			playsound(src.loc, 'sound/effects/ExplosionFirey.ogg', 100, 1)
+			qdel(src)
 
-		else if (istype(AM, /obj/critter/mouse) && (src.armed))
-			var/obj/critter/mouse/M = AM
+		else if (istype(AM, /mob/living/critter/small_animal/mouse) && !istype(AM, /mob/living/critter/small_animal/mouse/remy) && (src.armed))
```
[the remark this answers is not in the document]
Comment (Bartimeus973) [reply code/obj/item/mousetrap.dm]: The mighty have fallen. Remy is now mortal.

File: code/mob/living/critter/ai/generic_critter.dm [hunk 4]
```diff
@@ -191,7 +191,7 @@
 		var/obj/item/reagent_containers/food/snacks/T = holder.target
 		if(C && T && BOUNDS_DIST(holder.owner, holder.target) == 0)
 			holder.owner.set_dir(get_dir(holder.owner, holder.target))
-			T.Eat(C)
+			T.Eat(C, C, TRUE)
```
Comment (amylizzle) [inline code/mob/living/critter/ai/generic_critter.dm:194]: `by_matter_eater` being set to `TRUE` won't do anything unless the mob has a biolholder and the mattereater effect. See code/obj/item.dm:375

It should almost certainly be `FALSE` in the default critter AI, and overridden by a subclass if you want matter eater abilities on a mob.

File: code/mob/living/critter/ai/mouse.dm [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	New()
+		..()
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/mouse, list(src))
+
+/datum/aiTask/prioritizer/critter/mouse/New()
+	..()
+	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
+	transition_tasks += holder.get_instance(/datum/aiTask/sequence/goalbased/critter/eat, list(holder, src))
+
+/datum/aiHolder/mouse_mad
```
Comment (amylizzle) [inline code/mob/living/critter/ai/mouse.dm:11]: I kind of want this to be `/datum/aiHolder/mouse/mad`, ditto for the prioritizer. It doesn't matter, but it is a little cleaner.

File: code/mob/living/critter/small_animal.dm [hunk 6]
```diff
@@ -212,10 +219,48 @@ ABSTRACT_TYPE(/mob/living/critter/small_animal)
...
+		src.set_a_intent(INTENT_HELP)
+		return can_act(src,TRUE)
+
 /mob/living/critter/small_animal/mouse/weak
 	health_brute = 2
 	health_burn = 2
 
+/mob/living/critter/small_animal/mouse/mad
+	ai_type = /datum/aiHolder/mouse_mad
+	var/obj/machinery/wraith/rat_den/linked_den = null	//for mice spawned by plaguerat dens
```
Comment (amylizzle) [inline code/mob/living/critter/small_animal.dm:238]: probably better to keep this on the `ratden` subclass. No point having normal angry mice keeping an extra empty var.

File: code/obj/item/mousetrap.dm [hunk 7]
```diff
@@ -315,20 +315,27 @@
 			src.visible_message("<span class='alert'><b>[P] is caught in the trap and explodes violently into a rain of gibs!</b></span>")
 			P.gib()
 
-		else if ((ismobcritter(AM)) && (src.armed))
-			var/mob/living/critter/C = AM
-			src.triggered(C)
-			C.visible_message("<span class='alert'><B>[C] accidentally triggers the mousetrap.</B></span>",\
-				"<span class='alert'><B>You accidentally trigger the mousetrap!</B></span>")
+		else if (istype(AM, /mob/living/critter/small_animal/mouse/weak/mentor/admin)) //The admin mouse fears not your puny attempt to squish it.
```
Comment (amylizzle) [inline code/obj/item/mousetrap.dm:318]:
```suggestion
		else if (istype(AM, /mob/living/critter/small_animal/mouse/weak/mentor/admin) && src.armed) //The admin mouse fears not your puny attempt to squish it.
```

**Review by amylizzle — CHANGES_REQUESTED** [review]

File: code/mob/living/critter/ai/generic_critter.dm (hunk 4 above)
Comment (Bartimeus973) [reply code/mob/living/critter/ai/generic_critter.dm:194]: I thought the true here meant "bypass_utensils" from food_and_drink.dm? To allow critters to eat things such as soup, spaghetti and more?

Comment (amylizzle) [reply code/mob/living/critter/ai/generic_critter.dm:194]: You are absolutely right, I didn't realise `Eat()` overrides don't call parent.

File: code/mob/living/critter/ai/mouse.dm [hunk 8]
```diff
@@ -8,12 +8,12 @@
...
+/datum/aiHolder/mouse/mad
 	New()
 		..()
-		default_task = get_instance(/datum/aiTask/prioritizer/critter/mouse_mad, list(src))
+		default_task = get_instance(/datum/aiTask/prioritizer/critter/mouse/mad, list(src))
 
-/datum/aiTask/prioritizer/critter/mouse_mad/New()
+/datum/aiTask/prioritizer/critter/mouse/mad/New()
 	..()
 	transition_tasks += holder.get_instance(/datum/aiTask/timed/wander, list(holder, src))
```
Comment (amylizzle) [inline code/mob/living/critter/ai/mouse.dm:18]: the parent class already has this task, as well as eat. You just want to add attack here.

**Review by amylizzle — CHANGES_REQUESTED** [review]

File: code/mob/living/critter/ai/mouse.dm (hunk 8 above)
Comment (Bartimeus973) [reply code/mob/living/critter/ai/mouse.dm]: Donezo

**Review by amylizzle — APPROVED** [review]

**Comment by Bartimeus973** [comment]
Should work, but for reasons that confuse me greatly, spawning as a mentor/admin mice transforms it into a regular weak mice after a game tick.

**Comment by amylizzle** [comment]
> Should work, but for reasons that confuse me greatly, spawning as a mentor/admin mice transforms it into a regular weak mice after a game tick.

https://github.com/goonstation/goonstation/blob/17aa6b2804484833af2825d436e8af3dd2481fc9/code/mob/living/critter/small_animal.dm#L3049

**Review by frawhst — APPROVED** [review]
idk where the 'merge conflict' label is but this has some!!

**Comment by Bartimeus973** [comment]
Merge issues came from the roach mobification. This should fix it.

**Pull request merged**